mappings/net/minecraft/client/render/VertexConsumerHelper.mapping

```diff
@@ -0,0 +1,7 @@
+CLASS net/minecraft/class_4720 net/minecraft/client/render/VertexConsumerHelper
```

Comment:
Think this name needs to indicate that this consumer routes the output to two different consumers.  `Helper` doesn't convey that.

I called it `DualVertexConsumer`, but I dunno if that is best.  Maybe `DualOutputVertexConsumer` or `DuplicatingVertexConsumer`?

Reply:
This is the utility class. We use it like `VertexConsumerHelper.union()`. You don't say `Arrays.asList()`'s `Arrays` need to be `ArrayToList` so that `ArrayToList.asList()` makes sense to you

Reply:
It only does one thing and the class that handles the implementation is package-private.   `union` is also a bad name for the method, in that overall the names give more or less no indication of what this thing does or how it is used.

Reply:
from the param type of 2 vertex consumers and the return type of one vertex consumer and `union` name, you can clearly see that these two vertex consumers are unioned together. there is no sign of "duplicating" as the constuctor explicitly prevents duplicate consumers from passed into the argument. dual output maybe, but that is not clear either.

also, in #975, this class is also mapped as `net/minecraft/util/WoodType`. Feel free to vote on that.

Reply:
We clearly don't agree. But doesn't matter much. I can continue to use my own mappings.

Reply:
Yep, that is the advantage loom offered us over forgegradle (more flexible mappings!)

Reply:
I find `DuplicatingVertexConsumer` to be the clearest name proposed here, and if it is not used for the helper class I'd want to see it used for the implementation. And I think `VertexConsumerHelper#union` should be renamed as well, its behaviour is far from obvious.

Reply:
> the class that handles the implementation is package-private

At first sight I thought `VertexConsumerHelper` is package private or the utility method is; turns out only the impl class is, which does not matter. And this package-private class is not exposed in the return type of the static method either, which makes thing safe.

Mojang should have declared it private because that class should be. It is an impl detail and should not be leaked in any way, like `java.util.Arrays.ArrayList` is private as well.

Also this is not "duplicating" @Pyrofab: This is just same vertex data that should be transformed differently by a regular consumer and a color outline consumer or something like that from my previous experience, just like stream concat's result call the close of both base stream when closed isn't duplicating streams.

Reply:
I find your `Stream#concat` example ill-fitting. The main characteristic of a concatenated stream is to have the values of one stream coming after (concatenated with) the values of another. The main characteristic of `class_4589` is to send the vertex data to 2 delegate consumers. In doing so, it is effectively duplicating the data, or doing 2 operations at once (hence dual). Union is somewhat correct, but I have to look at the implementation to know what it actually means.
I was recommending renaming the implementation class because most modders look at a method's implementation before using it - `union() {return new Union()}` gives no additional information, whereas `union() {return new DuplicatingVertexConsumer()}` would make me understand very quickly what it is about.

Reply:
the impl class is already called union. comment on the new util classn i renamed it to vertex consumers.

Reply:
This isn't `concat` or `union`. This is very clearly mirroring output between render layers.

Might have to be something like `Layers.combine(layerOne, layerTwo);`

Also note that it can be chained:

```RenderLayer combined = Layers.combine(Layers.combine(layerOne, layerTwo), layerThree))```

Reply:
did you even read this @Sollace , this is freaking vertex consumers not freaking render layer util class! look at the stuff before you speak!

Reply:
I would rather not have unnecessary aggressive behaviour here. What Sollace said is approximately right, the combining vertex consumer is used to mirror output between 2 layers.